feat(crud todos): added collect todos

diegocoliveira opened the pull request:
* struct TodoCli
* struct Terminal
* struct Todos

diegocoliveira committed:
feat(crud todos): added collect todos
* struct TodoCli
* struct Terminal
* struct Todos

luisholanda commented on src/main.rs:
```diff
@@ -1,21 +1,121 @@
+    }
+
+    fn list(&mut self) -> Result<(), TerminalError> {
+        while let Some(id) = self.terminal.list_todos(self.todos.list())? {
+            let id = id.parse::<i32>().unwrap_or(0);
+            if !self.todos.exist(id) {
+                println!("{} Não existe um TODO com esse ID", Emoji("😕", ":/"));
+                self.terminal.press_key()?;
+                continue;
+            }
+            let action = self.terminal.ask_for_todo_action(id)?;
+            match action {
+                Action::Done(id) => self.done(id)?,
+                Action::Delete(id) => self.delete(id)?,
+                Action::Update(id, message) => self.update(id, message)?,
+                _ => (),
+            };
+            self.terminal.press_key()?;
+        }
+        Ok(())
```
Esse método ficou meio confuso, ele não deveria apenas listar os todos?

diegocoliveira replied:
A minha ideia de fluxo principal seria: 
1) adicionar todo
2) lista para possibilitar fazer alguma manipulação no todo

O que ficou confuso: o fluxo (lista para editar) ou o fato do método se chamar list e não verdade é outro submenu?

luisholanda commented on src/terminal.rs:
```diff
@@ -25,53 +34,153 @@ impl Display for TerminalError {
-    fn show_todo(&mut self, todo: &Todo) -> Result<(), TerminalError> {
+    pub fn list_todos(&self, list: Vec<&Todo>) -> Result<Option<String>, TerminalError> {
+        self.term.clear_screen().map_err(TerminalError::Stdout)?;
+        println!(
+            "################# {} ################# \n\n",
+            "LISTAGEM DOS TODOS"
+        );
```
Dado que não há nenhuma estilização ou coisa do genero em "LISTAGEM ...", podemos remover a formatação.

```suggestion
        println!("################# LISTAGEM DOS TODOS ################# \n\n");
```

diegocoliveira replied:
Na verdade, era para ter criado um método que recebia o valor do titulo e estilizar.... Mas a execução passou longe da ideia. Acabei dando só copia e colar . Vou ajeitar também.

luisholanda commented on src/main.rs:
```diff
@@ -1,21 +1,121 @@
+    todos: Todos,
+}
+
+impl TodoCli {
```
Nos métodos dessa struct, ficou um pouco confuso o uso de `terminal` para algumas coisas e `println` para outras. Não seria mais correto padronizar tudo estar no terminal?

diegocoliveira replied:
Na verdade até o próprio terminal tem local que uso println e outros writeln. Realmente está feio. Vou criar um método semelhante o input para imprimir todas as saídas (tanto interna quanto para chamadas externas)

luisholanda reviewed (commented):
A sua interface do `Todos` ficou muito boa, mandou bem!

luisholanda commented on src/todo.rs:
```diff
@@ -1,9 +1,68 @@
+
+pub struct Todos {
+    sequence: i32,
+    pub todos: BTreeMap<i32, Todo>,
```
BTW, idealmente esse cara deveria ser privado.

diegocoliveira replied:
putz... passei batido total nessa... Tanto que vou até verificar se eu acessei em algum lugar o BTW diretamente sem ser pelo métodos.

diegocoliveira committed:
refac: struct: terminal, TodoCli, Todos

1 - colocar BTreeMap privada
2 - criar método write no terminal e retirar todos println
3 - fazer estilização para os titulos
4 - ajustar método listar todo
5 - criar método editar todo

luisholanda reviewed (approved):
LGTM!
